src/unity/toolkits/activity_classification/sequence_iterator.cpp

```diff
@@ -24,27 +24,29 @@ static std::map<std::string,size_t> generate_column_index_map(const std::vector<
 
 /**
  *  Find the statistical mode (majority value) of a given vector.
- *  Based on https://en.wikipedia.org/wiki/Boyer–Moore_majority_vote_algorithm
  *
  * \param[in] input_vec a vector for which the mode will be calculated
 
  * \return    The most frequent value within the given vector.
  */
-static double vec_majority_value(const flex_vec& input_vec) {
-    size_t counter = 0;
-    double candidate = 0.0;
+
+static double vec_mode(const flex_vec& input_vec) {
+    std::vector<int> histogram;
     for (size_t i = 0; i < input_vec.size(); ++i) {
-        double value = input_vec[i];
-        if (counter == 0) {
-            candidate = value;
-            counter = 1;
-        } else if (candidate == value) {
-            counter++;
-        } else {
-            counter--;
+        int value = static_cast<int>(input_vec[i]);
+
+        assert(static_cast<double>(static_cast<double>(input_vec[i])) == input_vec[i]);
+
+        if(histogram.size() < (value + 1)){
+          histogram.resize(value + 1);
```

Comment:
It's a real shame that we have to potentially resize the vector multiple times on each call to `vec_mode`. This is the main way in which your version of this code is much slower than the old version!

It might be worth creating a single std::vector<int> instance and reusing it across all the calls to `vec_mode` inside a single call to `_activity_classifier_prepare_data_impl`. In the same file, you can see how there's one `curr_window_targets` vector being created in `_activity_classifier_prepare_data_impl` and e.g. passed by reference to `finalize_chunk`. Similarly, you could create a std::vector<int> (maybe called something like `histogram_buffer`) and pass it into `finalize_chunk` or `vec_mode` as appropriate.

For bonus points, note that we could determine the largest value of `input_vec` as it's being populated. (In practice, `input_vec` always seems to be `curr_window_targets`, which is just a cache of the class labels we've seen since the last chunk.) So you could even determine the correct size of the histogram vector up front.

Reply:
After giving this some more thought, although what I suggested would definitely be faster, it's not clear that it matters, given the other work going on in `_activity_classifier_prepare_data_impl`. And there's something to be said for keeping the code simpler. So feel free to take this comment as just an idea, could go either way. (The cast in the assertion should be fixed, though.)